### ISSUE TITLE
EngineSpec Refactor

### ISSUE THREAD

**Issue by frrist** [bacalhau-project/bacalhau#2351]
The gist of and motivation for this change is summarized in this [notion document]( https://www.notion.so/pl-strflt/Job-Schema-521ba6cdc06b4bdb940dbb151c576882)

The general idea is to replace the Specs Engine enum with the following:
```go
type Engine struct {
        // Type is a string representation of the Engines type
	Type string
        // Spec is the set of arguments(?) the engine will use during operation.
	Spec map[string]interface{}
}
```

Using a plain string for `Type` instead of dedicated enum type (as we do now) allows for the addition of new engines and submitting jobs at runtime without having to do a new release or approve the engine. For example, a compute provider wants to accept `spark_3.3` jobs. It should be able to do so just by publishing the engine in it's `NodeInfo` message. Users should be able to submit `spark_3.3` jobs without having to update their client to wait for the requester to update their server.

Using a `map[string]interface{}` for `Params` instead of a structure containing dedicated types (as we do now, e.g. `EngineSpecDocker`) enables the addition of new engines as their arguments will be dynamic to the engine `Type`. In future version we may consider making this a `[]byte`, but for now a `map` is more pragmatic.

### Acceptance Criteria
- [ ] Changes to `EngineSpecDocker` #2381 
  - [ ] move from the `modules` package to the `executor` package.
  - [ ] provide conversion between the `Engine.Param` and concrete `EngineSpecDocker` type
- [ ] Changes to `EngineSpecWasm` #2381
  - [ ] move from the `modules` package to the `executor` package.
  - [ ] rename `EntryPoint` to `Entrypoint`
  - [ ]  provide conversion between the `Engine.Param` and concrete `EngineSpecWasm` type
- [ ] Delete `JobSpecLanguage`. https://github.com/bacalhau-project/bacalhau/pull/2373
  - [ ] Delete its executor and related bits

**Comment by frrist** [comment]
In addition, we must determine whether we should permit the creation of ["invalid jobs"](https://github.com/bacalhau-project/bacalhau/blob/89c97a4a3087b4ac2db4c988a8bd48c7c626e3cb/pkg/model/engine.go#L21) through the Command Line Interface (CLI). At present, during job creation, we validate its engine against a list of supported engines. In principle, this change would eliminate the notion of "valid jobs," as it grants users the flexibility to specify any engine name of their choice. Instead of rejecting jobs with unrecognized engines, we could issue a warning to users about the utilization of an "unknown" engine.

For now I think it will be easiest to leave this validation in place, and remove it in a follow on when we are ready to make the network more open.

**Comment by simonwo** [comment]
Definitely support the idea! A few questions:


- Do we have plans about how to describe the schema of executor specs for use by the CLI? Atm one of our core design choices is having a CLI that is accessible to not-particularly-technical users. E.g. `bacalhau docker run` mirrors `docker run` very closely, and that is very executor specific, and I don't think we want to lose the ability to provide a very specific UX. Are we just going to leave the CLI as is for now and revisit this later?
- What is the motivation behind moving `EnvironmentVariables` into the generic `Spec`? It's not guaranteed that all executors have a meaningful way of using them. 
    - The fact that they are common between Docker and Wasm atm is more an artifact of our Wasm support being POSIX-like, but we could definitely support Wasm runtimes that don't have a concept of env vars in the future. 
    - My acid test for "is it too specific to our current executors" is imagining an `excel-2016` executor that can do something meaningful with spreadsheets (like run calculations) – env vars don't neccessarily make sense in this context.
    - So can we just leave them in the individual executor specs?
- What are we going to do about [`Spec.AllStorageSpecs()`](https://github.com/bacalhau-project/bacalhau/blob/fe125fb81c5e5e4124e8da3e04be2ea4c515bec6/pkg/model/job.go#L202-L203)? It is meant to return pointers to all the storage objects contained within the Spec, to allow them to be generically transformed. E.g. right now if a WASM blob or Docker container is submitted inline, it can be [saved to IPFS if it is too large](https://github.com/bacalhau-project/bacalhau/blob/fe125fb81c5e5e4124e8da3e04be2ea4c515bec6/pkg/requester/jobtransform/inline_spec_storage.go#L19). We also have open issues about doing the same with URLs, e.g. #1888. I don't think we want to go back to having a big bucket of storage specs, because we specifically refactored away from that due to it being difficult to use. 
- ~Do we know how are we going to handle executor specific transformations by the requester?~ Such as #2255? I guess we can just have executor-specific code there as we would have to currently.

**Comment by frrist** [comment]
> Are we just going to leave the CLI as is for now and revisit this later?

Yes that seems easiest at the moment. My thinking is we'll translate the job arguments from the CLI as they exist now into the `Params map[string]interface{}` field. The keys of the map will correspond to the field names in the specific job spec type e.g. `EngineSpecDocker`. Compute nodes supporting the engine type are then free to transform the `Params` map to a concrete type for execution.

The gist of the current change will be to store the CLI args in command specific variables rather than decode them directly into a job spec as we do now.

---------------

> What is the motivation behind moving `EnvironmentVariables` into the generic Spec? 

Motivation was initially stated [here](https://www.notion.so/pl-strflt/Job-Schema-521ba6cdc06b4bdb940dbb151c576882?pvs=4#7637802c426740e084f51dfe8247c0f0). However, after giving the doc a closer look I remembered [we both disagreed with the change](https://www.notion.so/Job-Schema-521ba6cdc06b4bdb940dbb151c576882?d=bcff5bf1c6d4469a99b8e37a852ac1e6#7c099293a76a4dc68684fa3e4f12edcf) to move them to `Spec`. 

> My acid test for "is it too specific to our current executors" is imagining an excel-2016 executor that can do something meaningful with spreadsheets (like run calculations) – env vars don't neccessarily make sense in this context.

In that specific case, assuming we _did_ move `EnvironmentVariables` to the `Spec`, they would just be null. However, that would mean Requester nodes need to supply compute nodes with the `Spec` in addition to the `EngineSpec`, which we do now, but may not be what we want in the future. 
(An aside: I worry users may not realize Job's (and their bits) are communicated over the network in plain text and that putting seneitive information in the `EnvironmentVariables` means leaking it to the world.)

@wdbaruni what do you think?

-------

> What are we going to do about [Spec.AllStorageSpecs()](https://github.com/bacalhau-project/bacalhau/blob/fe125fb81c5e5e4124e8da3e04be2ea4c515bec6/pkg/model/job.go#L202-L203)?

Since the method is utilized by Requester nodes, it causes the contents of an `EngineSpec` to be exposed/leaky as the Requester node needs to inspect it. At present, this is acceptable due to the limited and well-defined set of Engine types. However, once custom engines become supported, the Requester will not have the ability to inspect the contents.
Since the `JobSpecLanguage` engine is being removed, the `JobSpecWasm` is the only one needing inspection (the `JobSpecDocker` isn't inspected using this method). It wouldn't be hard to preserve the methods functionality for the time being. But, in general, having a Requester node modify job specifications before sending them to compute nodes is not ideal, and probably won't be scalable as the network expands.
I think we could either:
- reject jobs with inline data larger than 5kb (our current limit), which I think means limit the total size of the `EngineSpec` blob.
- perform the transformation on the client side since the client ought to be aware of the `EngineSpec` they are issuing.

Either of the above would allow this method to be removed, and reduce the responsibilities of the Requester node.

----

> Do we know how are we going to handle executor specific transformations by the requester? 

In the same vein as the above point, I think we should move away from request nodes modify the job spec (with in reason) before passing to the compute nodes.

**Comment by simonwo** [comment]
So I agree with the top two responses – thanks. 

However, we seem to be adopting this idea that the requester node is both completely executor agnostic and also does not/can not modify the user's submitted spec, and I think this is worthy of more inspection. 

We have a number of features today/planned that involve the requester modifying the spec in an executor-specific way. All of these features are supporting an easier UX, better network reliability, or reproducability. For example:

<details>
<summary>Examples of features we have/need</summary>
• #1628 – allowing the user to submit (large) storage blobs inline from their local environment which are then converted into (small) IPFS references. The use case is for thin clients (such as web browsers) to be able to include up to 10Mb of data with their request which is easier for them than spinning up a full IPFS node or pinning it to some external storage first. This is also intended to cover things like scripts or code that the user is developing locally but may be changing job-to-job. 

This is definitely used today for WASM blobs (at least 2-3Mb in size) and for sending Python code from Jupyter notebooks (often more than 5Kb), and is of interest for other edge applications (e.g. a webcam node can submit an image along with the job to process the image). This involves looking inside the spec and being able to identify storage specs.  

But we don't necessarily want blobs of this size traversing the network and we almost definitely don't want to be sending out blobs like this in a gossipy way. So the requester saving these blobs off to some appropriate storage is a feature for UX and network reliability. There isn't really a way to replace this by getting the client to do something, and limiting the size of inline data to 5kb is not meeting the requirement.

• #1888 – allowing the requester to cache URL downloads for concurrent jobs. This is a reproducability requirement – otherwise the data used in the job is not really captured in the job spec (assuming the general case of non-idempodent HTTP results). 

It is also a determinism requirement – the powerful principle that we should retain is that "the same job running concurrently over all nodes should be doing the same work at every node", which is the basis for determinisitic verification. Right now we allow different compute nodes to receive different URL inputs which in the general case makes it not possible to deterministically verify the results and spot bad actors. It's _also_ a safety requirement – we shouldn't allow people to DOS remote sites by specifying some large remote input and `--concurrency=1000`.

So there is a job for the requester here in downloading these inputs once, caching them accessibly (e.g. on IPFS), and then ensuring each compute node receives the same CIDs. Again, in general it involves being able to identify the StorageSpecs within each spec. Again, it would be poor UX and render the URL inputs flag useless to require the client to do this just to ensure sane computation.

• #2255 – allowing the requester to resolve a Docker image into a reproducible digest. This is a bug that our real users have already hit multiple times, where they make several pushes to `mytag:latest` and then are baffled as to why the previous version of the image is being used. It also has the same issues discussed above for URL inputs because different compute nodes will have different cache states, so if you are running a job concurrently you may end up with nodes running the job in different images = bad. Obvious issue for reproducability also in that being able to re-run some Docker job with a latest tag is not generally possible without knowing what the state of latest was when the job was run. 

We have already had some back and forth on this issue, but I've not heard any compelling arguments as to why the requester should not _be able_ to do this, _optionally_. I don't think telling the users "you should use a sha256 digest of the image, actually" is practical and again to require the client to do this would require them to have Docker tools installed on their machine/web browser, which again is not practical.

• unlogged yet, but for computation involving a TEE or Secure Enclave, someone needs to verify that the signatures delivered by the Enclave are actually valid before releasing encrypted data to it. Right now, the flow being proposed by Encloud fits well in the Bacalhau architecture as another executor, but it relies on the Requester node being a trusted agent for the user (something it already is) and being able to run executor-specific code to verify sigs. If we were to push this requirement onto the user, it would massively complicate client code (essentially clients would have to run a server with a key management service) and again make these jobs inaccessbile from constrained environments like browsers and smart contracts. 
</details>

Tl;dr:

- We need these features and more like them to ensure network reliability, scale, security and reproducability (all things the core team should be concerned with) and the requester node is in best place to have these responsibilities through virtue of it being a user-trusted node and the one that is orechestrating the job.
- Moving these responsibilities to the client either doesn't meet the requirement, would place too high a UX burden on users, or wouldn't be actionable by thin clients (web browsers, smart contracts, edge nodes).
- Some of these features used executor specific functionality (i.e. resolving a Docker tag) whereas others do not (e.g. validation, moving data from one type of storage to another). 

### Executor-aware requesters 
I completely understand that introducing generic engine types means that not all requester nodes will be able to carry out all of these functions. However, I think we would be missing a trick if we said that requester nodes _cannot_ carry out these functions _if they are able to_. I agree that people should not be _required_ to run a requester node with access to a Docker daemon, or in general requester nodes should not _need_ to know in detail about every engine they will orchestrate, but there is clearly extra value that can be delivered if they choose to. 

I don't see why we shoud not enable a higher quality of service if they want to do that. We have talked in the past about how there is a market-making opportunity in providing a higher quality of requester node service, or providing requester nodes with access to specific features like Secure Enclaves or HIPA-compliance, and we should be enabling that to happen. The analogy is with Filecoin-wrapper services like Estuary or Web3.Storage – these only exist to provide a more friendly, higher quality, managed service interface for things that people can already do with Lotus. Requester nodes or plugins to them can adopt the same role in our network. 

So **my proposal** is that we allow requesters to run executor specific features if they want to, but do not make it a requirement. We don't have any story about plugins yet, so for now that just means that we continue to ship all of the executors in a single binary, and if the requester detects that the executor is installed it can run executor-specific transformations, else it just doesn't do that. I do not think this has a problem with scale.

### Seeing inside the job spec
> However, once custom engines become supported, the Requester will not have the ability to inspect the contents.

I don't think this has to be true – I don't think we have to say "executor-specific parts of the spec are opaque". After all, we have schema languages that allow data to be described in a structured manner. And, in a way that can allow inspection of it at runtime by a client that has never seen that data before. 

As an example of what we could have instead of an opaque spec, we could instead specify:

```go
type Engine struct {
    Type CID
    Spec any
}
```

where the `Type` is a CID that resolves to an IPLD schema document that describes `Spec`. When a node (requester or even a client!) sees a engine `Type`, it can download the IPLD schema from IPFS and use to perform generic transformation of the job spec. The example I'm thinking of is those storage transformations - it can just walk the tree of the schema and identify any storage fields, and then transform them generically according to its policy.

This also allows the requester to validate the users submitted `Spec` without needing to have support built-in – it can just use the schema it downloads to validate it. This is actually better for orchestration, because it allows the requester to reject the job rather than having to send it out to N+ compute nodes and then receive simple validation errors from all of them.

Using the CID means we also get all the familiar advantages of content-addressing, i.e. compute nodes would advertise what CIDs they support, and then there can be no confusion about naming (e.g. my `python` is not the same as your `python`) and versioning (e.g. my spec is for a v2, but your compute node is only accepting v1 – they would have different CIDs).

We also know we're going to need a story for the CLI later – an accessible schema for the engine spec could well be a key part of that support.

As ever, if the requester cannot resolve the CID, it can just skip and validation and pass the spec through to supporting compute nodes anyway.

I would welcome thoughts on that approach.

> I think we should move away from request nodes modify the job spec (with in reason) before passing to the compute nodes.

Could you say a bit more about this? I've outlined a few features above where this is valuable. 

Are we worried about content addressing and/or reproducability? If so, I believe that if we transform the job spec, adding a field to refer to the CID (or equiv) of v1 of a job should answer most of our questions there – it would be clear to everyone on the network that the job spec being run has been transformed, and how to find/index the original. W.r.t octostore/amplify, it would still be possible to allow the user to query using a the job spec they submitted for the executions that were carried out as part of that transformed job.

**Comment by wdbaruni** [comment]
Regarding environment variables popping up to the spec level is that env variables should not be restricted to executor engines only, but we can expect storage source and publisher implementations that can accept env variables. So instead of having to define them in different places in the job spec, it might make sense to put them in one place?

However, if we believe we will move in a direction where a single job can have multiple tasks similar to IPVM invocation, then it make sense to have env variable at the task/executor level instead. 

I am neutral here and don't have strong preference towards any

**Comment by wdbaruni** [comment]
Going through the comments, I want to clarify that the main goal here is to _allow_ the requester to be agnostic of the executor engines to orchestrate jobs properly, and to allow introducing new engines with minimal code change. As Simon mentioned, it doesn't mean that some Requester Node implementations won't provide added value to the by inspecting the jobs if they wish to, specially when we talk about docker which I expect to always form 90% of the jobs in the network. 

In terms of job immutability and transforming the submitted job in the Requester side, today we are treating jobs as immutable _after_ they are created in Requester Node and not necessary what the user submits. I think this is fine for now, and we can always revisit this in the future by allowing the users to pass a flag to treat the submission as immutable, which in this case the Requester will fail if the submitted job is not compatible instead of trying to make it work

However, for the Requester Node that we are implementing and maintaining, we should really think twice before allowing the Requester to inspect and transform the job, as we don't want that to become the norm, I believe it can be a premature optimization now as we don't have a lot of customers yet, and can be difficult to disable it after few are using it. Basically I prefer to focus on building a higher quality core for our product with minimal dependencies and distractions, in favour of solving every single problem that users can solve on their own with minimal effort. 

The examples that were mentioned:
- https://github.com/bacalhau-project/bacalhau/pull/1628 – Caching large payloads in IPFS: Based on my understanding, newly added IPFS content can take time to be advertised to the network and it will depend whether the compute nodes have direct peering with the requester to know of new content immediately. It also adding a dependency on IPFS. What if node that can execute the job (e.g. through node labeling) doesn't support IPFS? Unless we had critical use cases we wanted to support, I would've preferred users solve this problem on their side. We can explore increasing the 5kB limit to accommodate more users. We can also explore compressing the gossipsub message by replacing the large payloads with their hashes or CIDs, but without having to publish to IPFS
- https://github.com/bacalhau-project/bacalhau/issues/1888 – allowing the requester to cache URL: Now we need to solve this for S3 content as well as the are also mutable. It should've been the user's responsibility to make sure the content doesn't change while the job is executing, or they can provide checksum for the content which allows failing the job if the checksum doesn't match the downloaded content, similar to what we do with S3 inputs
- https://github.com/bacalhau-project/bacalhau/issues/2255 – allowing the requester to resolve a Docker image into a reproducible digest: Wouldn't querying the registry in the compute nodes and refreshing the cache if stale solve this issue without adding a dependency on the requester node? It will still be the user's responsibility to make sure the image is not updated while submitting a job with high concurrency, which doesn't sound bad to me, specially that 99% of our jobs are with no concurrency. We can also improve the user experience by sharing the digest of the image that was used to run the job after the job is completed to help them debug incidents.

**Comment by frrist** [comment]
In a sense, given these requirements, the Requester Nodes seems similar to an Estuary Node, but for compute rather than storage. It provides extra utility for executing jobs on Bacalhau. Sure Clients could communicate directly with Compute nodes, but doing so via a Requester node provides a better UX/Higher quality of service to the Client.  

----------

An aspect I would like to clarify more precisely is the distinction between the protocols employed in the interactions between a Requester and a Compute Node, versus those between a Requester and a Client. I anticipate that the protocol governing the relationship between a Requester and a Client will be highly adaptable, considering the diverse services a Requester can offer to a Client. Conversely, the protocol for the connection between a Requester and a Compute Node is likely to be more rigid, as it pertains to crucial factors such as determinism, reproducibility, and security. My thinking here initially fell mostly into the latter, but this conversation has made clear the value of both. (When I say protocol here I mean the structures we are passing back and forth over a transport among nodes, not the transport specifically.)

> In terms of job immutability and transforming the submitted job in the Requester side, today we are treating jobs as immutable after they are created in Requester Node and not necessary what the user submits.

To this point, and my desire for more clarification above, might it be worth breaking apart protocol at the Client<->Requester layer vs Requester<->Compute layer? I think this is something worth considering, and seems like a path we are already heading down given this proposal to the EngineSpec.

-----------------

> the `Type` is a CID that resolves to an IPLD schema document that describes `Spec`. When a node (requester or even a client!) sees a engine `Type`, it can download the IPLD schema from IPFS and use to perform generic transformation of the job spec.

I am 100% on board with this design and am strongly in the camp of ""(Engine) Specifications should be self describing." And to @wdbaruni's point regarding "building a higher quality core for our product with minimal dependencies" I think this helps move us there by allowing the enhancement/support offered by the Requester to be "pluggable".

-----------------
> > I think we should move away from request nodes modify the job spec (with in reason) before passing to the compute nodes.
>
> Could you say a bit more about this? I've outlined a few features above where this is valuable.

I am contemplating this primarily from the perspective of protocol design. Assuming we permit Requester nodes to alter the state of a job specification submitted by a Client before forwarding it to a Compute node, I would prefer to have two distinct job specifications. One would be dedicated to communication between a Requester and a Client, while the other would be a more rigid specification designed for interactions between the Requester and the Compute node.

-----------

Some drive by comments on the github issues shared above:

- https://github.com/bacalhau-project/bacalhau/pull/1628: I think this implies we need to offer some kind of storage solution along side the Requester node. As @wdbaruni states I would prefer to place this responsibility on the Client.

- https://github.com/bacalhau-project/bacalhau/issues/2255: Possible unpopular opinion- we should disallow the usage of `latest` in the DockerEngine spec (and just in general). And this doesn't mean the Requester couldn't resolve this before running the job, but we should always be explicit about versioning, especially in the protocol, otherwise things will (probably) break or behave unpredictably.

**Comment by simonwo** [comment]
So I think this discussion is really about the two competing priorities of making our users successful in using our product and keeping our technical complexity at a manageable level. Happily I think we are all agreed that we need to be successful in both of these domains, and the discussion is just about how we balance those competing forces in order to achieve that dual success.

My perspective is that in all of the examples mentioned, we have added/planned the feature in direct response to/observation of our users failing to use our product. In the comments above I have tried to outline why I think that passing responsibility to the client won't lead to user success. (Regretfully, saying RTFM to the users doesn't seem acceptable in 2023! 😉) What am I missing? Do you think that putting these responsibilities on the user will be successful? Perhaps we have different ideas about who the users are now?

So I would much rather have a discussion about how we can solve these problems in the least technically complex way, rather than jusy saying "it's the users problem". I'm very happy to accept the currently proposed solutions have complexity and it would be good to discuss alternative ideas that manage that downwards.

In any case, as long as we are all agreed on having a self describing spec that would be inspectable by the requester, I don't think we need to continue this discussion here, and we can kick the can down the road a little of whether or not these features are kept.

@frrist, I think your observation about there being two protocols here is a useful one. We have indeed talked about having separate API and internal job specs before, but I guess we've never done the work to actually do something separate. Very happy to accept that once the job is in the requester<->compute format, it no longer is mutable. Most of the places it is interesting to mutate jobs seem to be in dealing with user input.

**Comment by frrist** [comment]
> In any case, as long as we are all agreed on having a self describing spec that would be inspectable by the requester, I don't think we need to continue this discussion here, and we can kick the can down the road a little of whether or not these features are kept.

Agreed on self describing spec. I will be moving forward with the (edit) changes proposed in OP. 

Will IPLD-ify the EngineSpec at some point in the future as we don't need badly right now, and I suspect this IPLD-ified version will closely resemble the final version used for exchange between compute and requester.

**Comment by wdbaruni** [comment]
> Regretfully, saying RTFM to the users doesn't seem acceptable in 2023! 
LOL .. Yes, we need to be user obsessed and improve the user experience of our solution for _most_ of our users. However, if we accommodate every single request, we will end up with an unmaintainable and complex system. Sometimes less is more! So we need to evaluate the long term implications of each feature, understand how many users are asking for it and why, and then decide if the value of introducing it outweigh the added complexity.

I do believe we are on the same page in making our system flexible and allow the requester to do work on behalf of the client to improve the user experience. My feedback is just to evaluate those features more thoroughly
